### PR TITLE
[Model] Add MiniMax-H3 support

### PR DESCRIPTION
**Pull request by nholmber**
# Add MiniMax-H3 Audio-Video Generation Support

## Summary

Adds xDiT support for MiniMax-H3 joint video and stereo-audio generation.

The integration supports FL2VA T2VA, first-frame I2VA, last-frame L2VA, first+last-frame FL2VA, and Ref2VA image references. It includes Ulysses, same-rank FSDP, Qwen3-VL tensor parallelism, fullgraph `torch.compile`, AITER BF16/FP8 attention, selective FP8/FP4 GEMMs, and quality-controlled audio-video encoding.

## Features

- **Generation modes:** T2VA, I2VA, L2VA, FL2VA, and Ref2VA image references
- **Joint output:** 124-frame video at 24 fps with synchronized 32 kHz stereo audio
- **Sequence parallelism:** Ulysses degrees 1, 2, 4, and 8
- **Padding correctness:** packed-varlen attention excludes alignment padding
- **FSDP:** Ulysses and FSDP can reuse the same ranks
- **Qwen3-VL TP:** text encoder TP reuses Ulysses ranks and saves 50.86 GiB per GPU at TP8
- **Attention:** AITER BF16 and gfx950 FP8 ASM attention
- **GEMMs:** selective torchao FP8 and AITER MXFP4 policies
- **Compilation:** full transformer `torch.compile(fullgraph=True)`
- **Encoding:** shared CRF-controlled H.264 + AAC encoder for MiniMax-H3 and LTX

## Usage

```bash
# Compiled single-GPU BF16 baseline
python -m xfuser.cli \
  --model MiniMaxAI/MiniMax-H3 \
  --prompt "A red fox trots through a snowy pine forest." \
  --height 544 --width 960 --num_frames 124 \
  --num_inference_steps 50 \
  --attention_backend AITER \
  --use_torch_compile

# U8 + Qwen3-VL TP8 + selective FP8 GEMMs + FP8 attention
python -m xfuser.cli \
  --model MiniMaxAI/MiniMax-H3 \
  --prompt "A red fox trots through a snowy pine forest." \
  --height 544 --width 960 --num_frames 124 \
  --num_inference_steps 50 \
  --ulysses_degree 8 --text_encoder_tp_degree 8 \
  --attention_backend AITER_FP8 --use_fp8_gemms \
  --use_torch_compile

# Fastest validated configuration: selective FP4 GEMMs + FP8 attention
python -m xfuser.cli \
  --model MiniMaxAI/MiniMax-H3 \
  --prompt "A red fox trots through a snowy pine forest." \
  --height 544 --width 960 --num_frames 124 \
  --num_inference_steps 50 \
  --ulysses_degree 8 --text_encoder_tp_degree 8 \
  --attention_backend AITER_FP8 --use_fp4_gemms \
  --use_torch_compile

# First-frame I2VA
python -m xfuser.cli \
  --model MiniMaxAI/MiniMax-H3 \
  --prompt "The subject begins to move." \
  --input_images first_frame.png

# Last-frame L2VA
python -m xfuser.cli \
  --model MiniMaxAI/MiniMax-H3 \
  --prompt "The scene resolves into the supplied image." \
  --task l2va --input_images last_frame.png

# First+last-frame FL2VA
python -m xfuser.cli \
  --model MiniMaxAI/MiniMax-H3 \
  --prompt "A continuous transition between both frames." \
  --input_images first_frame.png last_frame.png

# Ref2VA with ordered image references
python -m xfuser.cli \
  --model MiniMax-H3-Ref2VA \
  --prompt "Use the referenced character and visual style." \
  --input_images reference_1.png reference_2.png
```

`--text_encoder_tp_degree` reuses the model ranks and must be 1 or match the Ulysses degree. It does not multiply the process count.

https://github.com/user-attachments/assets/87163de1-58a2-4ea3-91fa-5fa8ada103df

## Performance

Tested on MI355X with:

- 960x544 output
- 124 frames
- 50 inference steps
- seed 42
- `torch.compile(fullgraph=True)`
- one three-step warmup
- median of measured iterations
- video/audio decode included
- MP4 encoding excluded

| Configuration | Median | Speedup vs U1 |
|---|---:|---:|
| U1 BF16 | 61.56s | 1.00x |
| U8 + Qwen TP8, BF16 | 13.45s | 4.58x |
| U8 + Qwen TP8, selective FP8 GEMMs | 12.26s | 5.02x |
| U8 + Qwen TP8, FP8 attention | 12.73s | 4.84x |
| U8 + Qwen TP8, selective FP8 GEMMs + FP8 attention | 11.48s | 5.36x |
| U8 + Qwen TP8, selective FP4 GEMMs | 11.33s | 5.43x |
| U8 + Qwen TP8, selective FP4 GEMMs + FP8 attention | **10.54s** | **5.84x** |

### Selective FP8 GEMMs

FP8:

- fused attention QKV
- fused SwiGLU gate/up

BF16:

- attention output
- MLP down
- AdaLN projection

### Selective FP4 GEMMs

FP4:

- fused attention QKV
- fused SwiGLU gate/up

FP8:

- attention output
- MLP down
- AdaLN projection

Full FP4 was faster but produced an unacceptable trajectory shift, so it is not used as the model default.

## Implementation Notes

- MiniMax-H3 packs text, video, and audio rows into one audiovisual sequence.
- The wrapper pads the packed sequence to 64 rows and passes live-key metadata to AITER varlen attention.
- Ulysses uses combined QKV all-to-all communication.
- FP8 attention dispatches the gfx950  `fmha_fwd_hd128_fp8_group_gfx950` ASM kernel.
- FP8 quantization and the AITER ASM call use `torch.library.custom_op`  boundaries to preserve FP8 dtypes under fullgraph compilation.
- Qwen3-VL TP uses replicated embeddings/norms, column-sharded QKV and gate/up,  and row-sharded attention output and MLP down projections.
- The unused Qwen language-model head is removed.
- Rank-0 text encoding with embedding broadcast remains available when text  encoder TP is disabled.

## Validation

- Tiny Diffusers transformer and xDiT wrapper match at U1.
- Fused QKV matches split Q/K/V output.
- U2 and U8 aligned and padded transformer outputs match the reference.
- Real padded FP8 varlen attention is finite and closely matches BF16.
- Fullgraph compile passes with packed BF16 and FP8 attention.
- Qwen3-VL TP8 output matches rank-0 encoding closely.
- FL2VA, Ref2VA, FP8 GEMM, FP4 GEMM, and FP8 attention examples were generated with synchronized audio.

## Known Limitations

- Ref2VA video and audio reference CLI inputs are not exposed yet.
- Video and audio VAE encoding/decoding are replicated; parallel VAE paths are not implemented.
- Qwen3-VL vision-tower and embedding parameters remain replicated under text encoder TP.
- PipeFusion is not implemented for MiniMax-H3.

### CODE REVIEW

File: xfuser/config/args.py [hunk 1]
```diff
@@ -514,6 +521,12 @@ def add_runner_args(parser: FlexibleArgumentParser):
             default=1,
             help="Tensor parallel degree.",
         )
+        parser.add_argument(
+            "--text_encoder_tp_degree",
```
Comment (avjves) [inline xfuser/config/args.py:535]: This should be added to the ModelCapabilities, so auto config validation will raise on models not supporting this.

File: xfuser/core/distributed/attention_backend.py [hunk 2]
```diff
@@ -830,47 +972,44 @@ def _aiter_fp8_attn_call(query, key, value, dropout_p, is_causal, attention_kwar
...
-    if AITER_FP8_HAS_DESCALE:
-        # If AITER_FP8_STATIC_SCALE_WITH_DESCALE is not set, use dynamic scaling.
-        # Set the environment variable XFUSER_AITER_FP8_STATIC_SCALE_WITH_DESCALE
-        # to a float value (i.e 2.5) to use static scaling.
-        if AITER_FP8_STATIC_SCALE_WITH_DESCALE is None:
-            scale = None
-        else:
-            scale=torch.tensor(AITER_FP8_STATIC_SCALE_WITH_DESCALE, dtype=torch.float32, device=query.device)
+    packed = _varlen_pack_keys(query, key, value, attention_kwargs)
+    if packed is not None:
```
Comment (avjves) [inline xfuser/core/distributed/attention_backend.py:976]: nit: why two separate if clauses with the same check?

**Review by avjves — APPROVED** [review]
Looks good! Could you still add the new te degree to ModelCapabilities?
There's now quite a lot of custom stuff in MiniMax, but it's nicely encapsulated within the model. Let's see in the future if some of that stuff should be more general if more similar models pop up.

File: xfuser/config/args.py (hunk 1 above)
Comment (nholmber) [reply xfuser/config/args.py:535]: Done. `ModelCapabilities` now declares `text_encoder_tp_degree=False` by default, and MiniMax-H3 explicitly enables it. The shared capability validation therefore rejects `--text_encoder_tp_degree > 1` for unsupported models. Added a regression test covering both the rejection and MiniMax-H3 capability.

File: xfuser/core/distributed/attention_backend.py (hunk 2 above)
Comment (nholmber) [reply xfuser/core/distributed/attention_backend.py:976]: Agreed; no reason to keep them separate. The packed tuple unpacking and varlen FP8 dispatch now share one `if packed is not None` block, with the dense path in the existing `else`.

**Review by avjves — APPROVED** [review]
